sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java

```diff
@@ -213,38 +211,39 @@ public MetricTrackingFnDataReceiver(
       labels.put(Labels.PCOLLECTION, pCollectionId);
+
+      // Collect the metric in a metric container which is not bound to the step name.
+      // This is required to count elements from impulse steps, which will produce elements outside
+      // of a pTransform context.
+      MetricsContainer unboundMetricContainer = metricsContainerRegistry.getUnboundContainer();
+
       MonitoringInfoMetricName elementCountMetricName =
           MonitoringInfoMetricName.named(MonitoringInfoConstants.Urns.ELEMENT_COUNT, labels);
-      this.elementCountCounter = LabeledMetrics.counter(elementCountMetricName);
+      this.unboundedElementCountCounter = unboundMetricContainer.getCounter(elementCountMetricName);
+
       MonitoringInfoMetricName sampledByteSizeMetricName =
           MonitoringInfoMetricName.named(Urns.SAMPLED_BYTE_SIZE, labels);
-      this.sampledByteSizeDistribution =
-          new SampleByteSizeDistribution<>(LabeledMetrics.distribution(sampledByteSizeMetricName));
-      this.coder = consumerAndMetadata.getValueCoder();
+      this.unboundSampledByteSizeDistribution =
+          new SampleByteSizeDistribution<>(
+              unboundMetricContainer.getDistribution(sampledByteSizeMetricName));
 
-      // Collect the metric in a metric container which is not bound to the step name.
-      // This is required to count elements from impulse steps, which will produce elements outside
-      // of a pTransform context.
-      this.unboundMetricContainer = metricsContainerRegistry.getUnboundContainer();
+      this.coder = consumerAndMetadata.getValueCoder();
     }
 
     @Override
     public void accept(WindowedValue<T> input) throws Exception {
-      try (Closeable close =
-          MetricsEnvironment.scopedMetricsContainer(this.unboundMetricContainer)) {
-        // Increment the counter for each window the element occurs in.
-        this.elementCountCounter.inc(input.getWindows().size());
-        // TODO(BEAM-11879): Consider updating size per window when we have window optimization.
-        this.sampledByteSizeDistribution.tryUpdate(input.getValue(), this.coder);
-        // Wrap the consumer with extra logic to set the metric container with the appropriate
-        // PTransform context. This ensures that user metrics obtain the pTransform ID when they are
-        // created. Also use the ExecutionStateTracker and enter an appropriate state to track the
-        // Process Bundle Execution time metric.
-        MetricsContainerImpl container = metricsContainerRegistry.getContainer(pTransformId);
-        try (Closeable closeable = MetricsEnvironment.scopedMetricsContainer(container)) {
-          try (Closeable trackerCloseable = stateTracker.enterState(state)) {
-            this.delegate.accept(input);
-          }
+      // Increment the counter for each window the element occurs in.
+      this.unboundedElementCountCounter.inc(input.getWindows().size());
```

Comment:
my reasoning on why this was a no-op:
- the scoped metric containers stuffs in thread-local and delegating counter pulls it out
- the scoping is not transitive, only the last scope matters
- we are now caching the unboundedMetricContainer.getCounter() result instead of recalculating it every accept.  However the contaner appears to reset counters but not remove them.

I was hoping existing tests would cover correctness since I'm not familiar with this code.

@kileys can you take a look? Thanks